File: backend/app/schemas/content.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import date, datetime

# Content
class Content(BaseModel):
    id: int
    title: str
    type: str
    overview: Optional[str]
    poster: Optional[str]
    backdrop: Optional[str]
    release_date: Optional[date]
    year: Optional[int]
    runtime: Optional[int]
    language: Optional[str]
    age_rating: Optional[str]
    age_rating_region: Optional[str] = None
    age_rating_source: Optional[str] = None
    age_rating_system: Optional[str] = None


class PaginatedContentResponse(BaseModel):
    items: List[Content]
    total: int
    limit: int
    offset: int


# Platform
class Platform(BaseModel):
    name: str
    availability_type: str
    platform_type: Optional[str] = None
    region_code: Optional[str] = None
    source_name: Optional[str] = None
    source_provider_id: Optional[str] = None
    display_priority: Optional[int] = None


# Rating
class RatingSourceItem(BaseModel):
    source_name: str
    display_name: str
    source_category: str
    raw_score: Optional[float] = None
    raw_score_scale: Optional[float] = None
    normalized_score: Optional[float] = None
    vote_count: Optional[int] = None
    rating_count_label: Optional[str] = None
    rating_url: Optional[str] = None
    fetched_at: Optional[datetime] = None
    included_in_unified_score: bool = False


class RatingsResponse(BaseModel):
    unified_score: Optional[int] = None
    source_count: int
    scoring_source_count: int = 0
    sources: List[RatingSourceItem]


# Summary
class Summary(BaseModel):
    unified_score: Optional[float]
    critic_score: Optional[float]
    audience_score: Optional[float]
    review_summary: Optional[str]
    pros: Optional[str]
    cons: Optional[str]
    verdict: Optional[str]


class InsightSummarySignal(BaseModel):
    label: str
    value: str


class InsightSummary(BaseModel):
    headline: Optional[str] = None
    summary: Optional[str] = None
    best_for: List[str]
    key_signals: List[InsightSummarySignal]
    watch_note: Optional[str] = None
    generated_from: List[str]
    confidence: str


class WatchProfileSchema(BaseModel):
    watch_feel: Optional[str] = None
    chips: List[str]
    best_for: List[str]
    consider_first: List[str]


class DecisionSupportSchema(BaseModel):
    headline: Optional[str] = None
    reasons: List[str]
    cautions: List[str]


class DecisionDisplayProfileSchema(BaseModel):
    identity: List[str]
    themes: List[str]
    feel: List[str]
    pace: Optional[str] = None
    best_for: List[str]
    consider_first: List[str]


class DecisionDisplayFactSchema(BaseModel):
    label: str
    value: str


class DecisionDisplaySchema(BaseModel):
    primary_insight: Optional[str] = None
    profile: DecisionDisplayProfileSchema
    supporting_facts: List[DecisionDisplayFactSchema]


class SourceSignalQualitySchema(BaseModel):
    storage_ready: bool
    frontend_ready: bool
    has_watch_guidance: bool
    has_source_signals: bool


class ContentDecisionLayerSchema(BaseModel):
    watch_profile: WatchProfileSchema
    decision_support: DecisionSupportSchema
    display: Optional[DecisionDisplaySchema] = None
    signal_quality: SourceSignalQualitySchema


class SeriesMetadata(BaseModel):
    number_of_seasons: Optional[int] = None
    number_of_episodes: Optional[int] = None
    series_status: Optional[str] = None
    series_status_normalized: Optional[str] = None
    in_production: Optional[bool] = None
    first_air_date: Optional[date] = None
    last_air_date: Optional[date] = None
    last_episode_air_date: Optional[date] = None
    next_episode_air_date: Optional[date] = None
    series_type: Optional[str] = None
    released_seasons_count: Optional[int] = None
    announced_seasons_count: Optional[int] = None
    next_season_number: Optional[int] = None
    next_season_air_date: Optional[date] = None
    next_season_year: Optional[int] = None
    has_announced_season: Optional[bool] = None
    season_summary_note: Optional[str] = None


class Genre(BaseModel):
    id: int
    name: str


class PlatformMetadata(BaseModel):
    id: int
    name: str
    platform_type: str


class CastCredit(BaseModel):
    person_id: int
    name: str
    character_name: Optional[str] = None
    profile_url: Optional[str] = None
    known_for_department: Optional[str] = None
    display_order: Optional[int] = None


class CrewCredit(BaseModel):
    person_id: int
    name: str
    profile_url: Optional[str] = None
    known_for_department: Optional[str] = None
    job: Optional[str] = None
    department: Optional[str] = None
    role_type: Optional[str] = None
    display_order: Optional[int] = None


class ContentCreditsResponse(BaseModel):
    content_id: int
    cast: List[CastCredit]
    directors: List[CrewCredit]
    creators: List[CrewCredit]
    crew: List[CrewCredit]

    
# Final Response Model
class ContentDetailsResponse(BaseModel):
    content: Content
    genres: List[str]
    platforms: List[Platform]
    ratings: RatingsResponse
    series_metadata: Optional[SeriesMetadata] = None
    insight_summary: InsightSummary
    decision_layer: Optional[ContentDecisionLayerSchema] = None
    summary: Optional[Summary] = None

File: frontend/types/content.ts
export type Content = {
  id: number;
  title: string;
  original_title?: string | null;
  type: "movie" | "series";
  overview?: string | null;
  poster?: string | null;
  backdrop?: string | null;
  release_date?: string | null;
  year?: number | null;
  runtime?: number | null;
  language?: string | null;
  original_language?: string | null;
  original_language_name?: string | null;
  age_rating?: string | null;
  age_rating_region?: string | null;
  age_rating_source?: string | null;
  age_rating_system?: string | null;
  unified_score?: number | null;
  source_count?: number | null;
  scoring_source_count?: number | null;
};

export type SeriesMetadata = {
  number_of_seasons?: number | null;
  number_of_episodes?: number | null;
  series_status?: string | null;
  series_status_normalized?: string | null;
  in_production?: boolean | null;
  first_air_date?: string | null;
  last_air_date?: string | null;
  last_episode_air_date?: string | null;
  next_episode_air_date?: string | null;
  series_type?: string | null;
  released_seasons_count?: number | null;
  announced_seasons_count?: number | null;
  next_season_number?: number | null;
  next_season_air_date?: string | null;
  next_season_year?: number | null;
  has_announced_season?: boolean | null;
  season_summary_note?: string | null;
};

export type PaginatedContentResponse = {
  items: Content[];
  total: number;
  limit: number;
  offset: number;
};

export type HomeQuickFilter = {
  label: string;
  filter_key: string;
};

export type HomeHero = {
  title: string;
  subtitle: string;
  quick_filters: HomeQuickFilter[];
};

export type HomeContentCard = {
  id: number;
  title: string;
  content_type: Content["type"];
  year?: number | null;
  poster_url?: string | null;
  backdrop_url?: string | null;
  runtime?: number | null;
  age_rating?: string | null;
  release_date?: string | null;
  unified_score?: number | null;
  source_count?: number | null;
  scoring_source_count?: number | null;
  primary_platform?: string | null;
  platforms: string[];
  decision_reason: string;
  chips: string[];
};

export type HomeBucket = {
  bucket_id: string;
  label: string;
  subtitle: string;
  refresh_strategy?: string | null;
  refresh_cadence?: string | null;
  items: HomeContentCard[];
};

export type HomeSection = {
  section_id: string;
  title: string;
  subtitle: string;
  section_type: "poster_rail" | "bucketed_rail" | string;
  refresh_strategy: string;
  refresh_cadence: string;
  items?: HomeContentCard[] | null;
  buckets?: HomeBucket[] | null;
};

export type HomeResponse = {
  hero: HomeHero;
  sections: HomeSection[];
  generated_for?: string | null;
  refresh_note?: string | null;
};

export type Genre = {
  id: number;
  name: string;
};

export type PlatformType = "ott" | "rating_source" | "review_source";

export type PlatformMetadata = {
  id: number;
  name: string;
  platform_type: PlatformType | string;
};

export type DiscoverContentParams = {
  content_type?: Content["type"];
  genre?: string;
  platform?: string;
  availability_type?: "streaming" | "rent" | "buy";
  sort_by?: "recent" | "top_rated";
  limit?: number;
  offset?: number;
};

export type PlatformAvailability = {
  name: string;
  availability_type: "streaming" | "rent" | "buy" | string;
  platform_type?: PlatformType | string | null;
  region_code?: string | null;
  source_name?: string | null;
  source_provider_id?: string | null;
  display_priority?: number | null;
};

export type RatingSourceItem = {
  source_name: string;
  display_name: string;
  source_category: "audience" | "critic" | "theatrical" | "internal" | string;
  raw_score?: number | null;
  raw_score_scale?: number | null;
  normalized_score?: number | null;
  vote_count?: number | null;
  rating_count_label?: string | null;
  rating_url?: string | null;
  fetched_at?: string | null;
  included_in_unified_score?: boolean;
};

export type RatingsResponse = {
  unified_score?: number | null;
  source_count: number;
  scoring_source_count?: number;
  sources: RatingSourceItem[];
};

export type Summary = {
  unified_score?: number | null;
  critic_score?: number | null;
  audience_score?: number | null;
  review_summary?: string | null;
  pros?: string | null;
  cons?: string | null;
  verdict?: string | null;
};

export type InsightSummarySignal = {
  label: string;
  value: string;
};

export type InsightSummary = {
  headline?: string | null;
  summary?: string | null;
  best_for: string[];
  key_signals: InsightSummarySignal[];
  watch_note?: string | null;
  generated_from: string[];
  confidence: "low" | "medium" | "high" | string;
};

export type DecisionDisplayFact = {
  label: string;
  value: string;
};

export type DecisionDisplayProfile = {
  identity?: string[];
  themes?: string[];
  feel?: string[];
  pace?: string | null;
  best_for?: string[];
  consider_first?: string[];
};

export type DecisionDisplay = {
  primary_insight?: string | null;
  profile?: DecisionDisplayProfile | null;
  supporting_facts?: DecisionDisplayFact[];
};

export type DecisionLayer = {
  display?: DecisionDisplay | null;
  watch_profile?: unknown;
  decision_support?: unknown;
  signal_quality?: unknown;
};

export type ContentDetailsResponse = {
  content: Content;
  genres: string[];
  platforms: PlatformAvailability[];
  ratings: RatingsResponse;
  series_metadata?: SeriesMetadata | null;
  insight_summary: InsightSummary;
  decision_layer?: DecisionLayer | null;
  summary?: Summary | null;
};

export type UserContentActionRequest = {
  user_id: number;
  content_id: number;
};

export type UserContentActionResponse = {
  message: string;
};

export type UserContentItem = Content;

export type WatchStatus = "none" | "watch_later" | "watched";
